### ISSUE TITLE
Define 1.0 release

### ISSUE THREAD

**Issue by Fernerkundung** [sentinelsat/sentinelsat#108]
I would like to get a discussion going what we define as the 1.0 milestone for `sentinelsat`.

I would argue that we are pretty much feature complete now - if this can be said about a project without a clear definition of expected features. As long as there are no dramatic changes to the Open Access Hub Software new additions will mainly be when new satellites/sensors are included, for which we are well prepared with the generalized layout of the API.

What I see mostly happening now is refinement and refactoring, where the biggest part is likely the refactoring of the download function to alleviate the `curl` dependency as our only hard dependency on a compiled library (and the biggest source of install/user issues).

What are changes you would expect to be made regarding a 1.0 release? I'll start with:

- [x] refactoring download function and remove `curl` dependency

**Comment by valgur** [comment]
I agree, it's a very good idea to have a concrete roadmap.

I think the most important and quite obvious goal for version 1.0 should be the stability of the API we provide. That is we should try to fix the API we have and only change it in a backwards-compatible manner or by explicitly adding deprecation warnings where that is not possible. This means mostly that any changes that would impact the API should be done before version 1.0. Using [semantic versioning](http://semver.org/) from then on would be a good idea to make life easier for users.

In terms of functionality I think the major goal, at least for the Python API, would be to make the server's API functionality available as client-side functions. I think we have that pretty much covered. The only thing I can think of that could be potentially useful and worth adding is accessing specific files from the zip archives directly, e.g. [this S2 manifest.safe file](https://scihub.copernicus.eu/dhus/odata/v1/Products('6732e4f7-d590-4d86-a605-cd91326ee0bb')/Nodes('S2A_MSIL1C_20170425T193901_N0204_R042_T13XDF_20170425T193904.SAFE')/Nodes('manifest.safe')/$value). It would not be particularly difficult to add, but I'm not sure how big a demand there would be for this.

Now, for the things I think we *should* consider doing.
- [ ] Get rid of the `pycurl` dependency: I think I have a working version of this on a personal branch, I just need to add some tests and submit a PR.
- [ ] Making our exceptions more specific. I would split `SentinelAPIError` into a couple of more specific ones:

  - `SentinelAPIError`: for relaying API responses that return an exception, e.g. for invalid query parameters. The response is expected to contain the `cause-message` in that case (as in #107).
  - `UnauthorizedError`: for cases where invalid credentials were provided or when the server does not accept them yet for whatever reason (a common issue for new users, I think).
  - `ServerError`: for cases when the server is unreachable, under maintenance or the response is just unparsable garbage.

- [ ] Adding automatic retries to all functions that interact with the server. The retry parameters would be configurable, perhaps as `SentinelAPI` properties. Something along the lines of [this](https://powerfulpython.com/blog/making-unreliable-apis-reliable-with-python/), except we probably can't use decorators.
- [ ] I would make checksumming downloaded files the default behavior.
- [ ] Add client-side product sorting.

Another minor idea: add an option to `query()` and `search` on the CLI to translate the OpenSearch attribute names to the ones the OData API uses (see [the example in the docs](https://github.com/valgur/sentinelsat/blob/api-improvements/docs/api.rst#getting-product-metadata)), which I find to be a lot nicer.

**Comment by Fernerkundung** [comment]
I would also like to add

- [ ] define a governance structure
- [ ] add CONTRIBUITE.rst with contribution guidelines

This is regarding the owner/maintainer/contributor structure. Currently everything has to go through @marcelwille as the repository owner. He stated that, this being inside a organisation account, that he won't be able to add maintainers to the repo. Without maintainer status we will however not be able to close issues, add tags to issues, etc. I think going forward we need to talk about this, since I feel the maintenance of `sentinelsat` might be considerably slowed down or puts a lot of pressure/work on @willemarcel.

**Comment by j08lue** [comment]
> The only thing I can think of that could be potentially useful and worth adding is accessing files directly from the zip archives, e.g. this S2 manifest.safe file.
@willemarcel, what do you mean by that? I have some libraries for parsing and sorting downloaded data by their meta data - but that is not what you were thinking of, I guess?

**Comment by j08lue** [comment]
> This is regarding the owner/maintainer/contributor structure. Currently everything has to go through @willemarcel as the repository owner. He stated that, this being inside a organisation account, that he won't be able to add maintainers to the repo. Without maintainers we will however not be able to close issues, add tags to issues, etc. I think going forward we need to talk about this, since I feel the maintenance of sentinelsat might be considerably slowed down or puts a lot of pressure/work on @willemarcel.

Maybe we should actually start with reorganizing the governance to be able to make realistic road maps for all the very good points points you are listing?

**Comment by Fernerkundung** [comment]
@j08lue That's a good idea - are you OK with that @willemarcel?

I guess to adapt any structure from other OSS projects we will need to fork the repo and [move the existing issues into the new repo](https://github.com/google/github-issue-mover). Or is there a way to transfer the ownership of a repo.

Anybody know a best practices guide for OSS project governance structures? I see a lot of projects using organization accounts and then have some sort of maintainer/contributer hierarchy.

The Documentation endpoint is something I can move, @willemarcel would need to change the PyPi endpoint and also upload new releases unless he'd be willing to transfer the ownership of that to someone or a shared account.

**Comment by willemarcel** [comment]
Yes, @Fernerkundung we can create an organization to `sentinelsat`. Sorry for the delay. The last weeks were very hard.

**Comment by Fernerkundung** [comment]
@willemarcel No need to appologize. We are all doing this in our freetime - and moving to a shared responsibility model hopefully helps us not putting too much stress on anybody :)

**Comment by j08lue** [comment]
> we can create an organization to sentinelsat

Yes, that is the way to go, @willemarcel. And I agree - no need to apologize. 

> Or is there a way to transfer the ownership of a repo.

Yes, there is, in the settings section. Create the organization first and then @willemarcel can transfer to that including issues and all.

**Comment by Fernerkundung** [comment]
> Create the organization first and then @willemarcel can transfer to that including issues and all.

I took the liberty to create an organization and invited all of you into it. https://github.com/sentinelsat

I propose to make @willemarcel @j08lue, @valgur and @Fernerkundung maintainers of the repo once it is transferred. Additionally I would propose to adopt an open meritocracy governance structures (see [http://oss-watch.ac.uk/resources/governancemodels](http://oss-watch.ac.uk/resources/governancemodels), which would mean that we invite contributions by anyone but 1 or 2 maintainers have to review the code before it is merged into master. Due to sentinelsat being a rather small project I would propose that the review of 1 maintainer is enough.

**Comment by willemarcel** [comment]
@Fernerkundung I transfered the repository and I agree with the rules. What is your username in Pypi?

**Comment by Fernerkundung** [comment]
@willemarcel that would be the same username as here - "Fernerkundung".

**Comment by willemarcel** [comment]
@Fernerkundung I gave you permissions to publish new versions

**Comment by valgur** [comment]
Thanks for embracing a truly open approach to open source, @willemarcel. :) I hope this relieves a lot of the maintenance pressure for you in a good way.

@j08lue What I had in mind was that you can access the contents of the zip files directly via the OAuth API. Pay attention to the URLs. For example:
* You can access the product's manifest.safe file https://scihub.copernicus.eu/dhus/odata/v1/Products('6732e4f7-d590-4d86-a605-cd91326ee0bb')/Nodes('S2A_MSIL1C_20170425T193901_N0204_R042_T13XDF_20170425T193904.SAFE')/Nodes('manifest.safe')/$value
* get a list of files and directories in that zip https://scihub.copernicus.eu/dhus/odata/v1/Products('6732e4f7-d590-4d86-a605-cd91326ee0bb')/Nodes('S2A_MSIL1C_20170425T193901_N0204_R042_T13XDF_20170425T193904.SAFE')/Nodes
* download a quicklook file or a quality mask https://scihub.copernicus.eu/dhus/odata/v1/Products('6732e4f7-d590-4d86-a605-cd91326ee0bb')/Nodes('S2A_MSIL1C_20170425T193901_N0204_R042_T13XDF_20170425T193904.SAFE')/Nodes('GRANULE')/Nodes('L1C_T13XDF_A009620_20170425T193904')/Nodes('QI_DATA')/Nodes('T13XDF_20170425T193901_PVI.jp2')/$value and so on.
I can imagine having access to files like these could be useful to someone in some situation. There is a more direct way to access most of that information though via the API (e.g. for the [quicklook](https://scihub.copernicus.eu/dhus/odata/v1/Products('6732e4f7-d590-4d86-a605-cd91326ee0bb')/Products('Quicklook')/$value) file).

**Comment by j08lue** [comment]
That is pretty cool, I did not know of that functionality. Perhaps you could also use it to pass a band file URL through some stream directly to `rasterio.open` and open it off the server [like with AWS resources and much else](https://github.com/mapbox/rasterio/issues/977#issue-205238864)... 

But we should not make our little system too complicated. This could also be a separate module or even package.

Let us open a separate issue on this and discuss it there.

**Comment by valgur** [comment]
Any new ideas that have been suggested in this thread should be converted to a separate issue, so it's easier to discuss them and keep track of via our milestone labels. I think I took care of that for the ideas presented so far, but feel free to create any new issues if I missed something.

**Comment by Fernerkundung** [comment]
All proposed changes have been moved to separate issues.
